src/Rule/Hash.php

```diff
@@ -0,0 +1,101 @@
+<?php
+/**
+ * Particle.
+ *
+ * @link      http://github.com/particle-php for the canonical source repository
+ * @copyright Copyright (c) 2005-2015 Particle (http://particle-php.com)
+ * @license   https://github.com/particle-php/validator/blob/master/LICENSE New BSD License
+ */
+namespace Particle\Validator\Rule;
+
+use Particle\Validator\Rule;
+
+/**
+ * This rule is for validating if a value is a valid cryptographic hash.
+ *
+ * @package Particle\Validator\Rule
+ */
+class Hash extends Rule
+{
+    const ALGO_MD5 = 'md5';
+    const ALGO_SHA1 = 'sha1';
+    const ALGO_SHA256 = 'sha256';
+    const ALGO_SHA512 = 'sha512';
+    const ALGO_CRC32 = 'crc32';
+
+    /**
+     * A constant that will be used when the value is not a valid cryptographic hash.
+     */
+    const INVALID_FORMAT = 'Hash::INVALID_FORMAT';
+
+    /**
+     * The message templates which can be returned by this validator.
+     *
+     * @var array
+     */
+    protected $messageTemplates = [
+        self::INVALID_FORMAT => '{{ name }} must be a valid hash'
+    ];
+
+    /**
+     * @var string
+     */
+    protected $hashAlgorithm;
+
+    /**
+     * @var bool
+     */
+    protected $allowUppercase;
+
+    /**
+     * Construct the Hash validator.
+     *
+     * @param string $hashAlgorithm
+     * @param bool $allowUppercase
+     */
+    public function __construct($hashAlgorithm, $allowUppercase = false)
+    {
+        $this->hashAlgorithm = $hashAlgorithm;
+        $this->allowUppercase = $allowUppercase;
+
+        $this->messageTemplates = [
+            self::INVALID_FORMAT => sprintf('{{ name }} must be a valid %s hash', $hashAlgorithm)
+        ];
+    }
+
+    /**
+     * Validates if the value is a valid cryptographic hash.
+     *
+     * @param mixed $value
+     * @return bool
+     */
+    public function validate($value)
+    {
+        if ($this->hashAlgorithm == self::ALGO_MD5 && $this->validateHexString($value, 32)) {
+            return true;
+        } elseif ($this->hashAlgorithm == self::ALGO_SHA1 && $this->validateHexString($value, 40)) {
+            return true;
+        } elseif ($this->hashAlgorithm == self::ALGO_SHA256 && $this->validateHexString($value, 64)) {
+            return true;
+        } elseif ($this->hashAlgorithm == self::ALGO_SHA512 && $this->validateHexString($value, 128)) {
+            return true;
+        } elseif ($this->hashAlgorithm == self::ALGO_CRC32 && $this->validateHexString($value, 8)) {
```

Comment:
https://scrutinizer-ci.com/g/particle-php/Validator/inspections/7b95f690-1a26-496c-bbac-886cca2fea25/code-structure/operation/Particle%5CValidator%5CRule%5CHash%3A%3Avalidate?expandCoverage=1

I'm not happy with the looks of the `elseif` structure here, and I believe this could be improved using a switch or an array. I will look into this later today :)